### PR TITLE
Feature/better error management

### PR DESCRIPTION
**Pull request by StarpTech**
This is a first draft of introducing error codes to fastify. It only handles the codes for the contentTypeParser.

# TODO

* [x] Create error codes for all others
* [x] Document the error codes

Previous PR: https://github.com/fastify/fastify/pull/957

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/custom-parser.test.js [hunk 1]
```diff
@@ -796,10 +807,12 @@ test('Should allow defining the bodyLimit per parser', t => {
       }
     }, (err, response, body) => {
       t.error(err)
+      const expectedError = new FST_ERR_CTP_BODY_TOO_LARGE()
       t.strictDeepEqual(JSON.parse(body.toString()), {
-        statusCode: 413,
+        statusCode: expectedError.statusCode,
+        code: expectedError.code,
```
Comment (mcollina) [inline test/custom-parser.test.js]: I think we should keep those explicit in the tests

File: test/custom-parser.test.js [hunk 2]
```diff
@@ -312,7 +323,7 @@ test('the content type should be a string', t => {
     fastify.addContentTypeParser(null, () => {})
     t.fail()
   } catch (err) {
-    t.is(err.message, 'The content type should be a string')
+    t.is(err.message, new FST_ERR_CTP_INVALID_TYPE().message)
```
Comment (mcollina) [inline test/custom-parser.test.js]: I would prefer for the message to be explicit here.

Comment (StarpTech) [reply test/custom-parser.test.js]: What's the benefit? I think the error name is already very explicit. In that way we can avoid modifying all tests when we change something in the future.

Comment (mcollina) [reply test/custom-parser.test.js]: Sure, but I would love to avoid testing all errors individually.

**Comment by mcollina** [comment]
@DustinDeus any updates here?

**Comment by StarpTech** [comment]
I will work on it in smaller updates I'm too busy right now. Any help is welcome.

**Comment by cemremengu** [comment]
@StarpTech happy to help, what's left to do can you provide some kind of check list?

**Comment by StarpTech** [comment]
@cemremengu sure

* Have a look at how we want to use errors in the future.
* Address @mcollina point
* Change it everywhere & update tests
* Document the new errors

**Comment by cemremengu** [comment]
@StarpTech Reverted test to use string messages. It would look like this:

```js
      t.strictDeepEqual(JSON.parse(body.toString()), {
        statusCode: 413,
        code: "FST_ERR_CTP_BODY_TOO_LARGE",
        error: 'Payload Too Large',
        message: 'FST_ERR_CTP_BODY_TOO_LARGE: Request body is too large'
      })
```

Should I push to your branch with your permission? How should we progress?

**Comment by cemremengu** [comment]
I think there is problem with `schema.js`. The test below fails:

```js
test('Should throw of the schema does not exists', t => {
  t.plan(1)
  const fastify = Fastify()

  fastify.route({
    method: 'GET',
    url: '/:id',
    schema: {
      params: 'test#'
    },
    handler: (req, reply) => {
      reply.send(typeof req.params.id)
    }
  })

  fastify.ready(err => {
    t.is(err.message, 'Schema with id \'test\' does not exist!')
  })
})
```

If I change/delete the `error.code` below (since we now have string codes like FST....) in `shared-schemas.test.js`, ajv will complain that schema must be a an object or buffer

```js
Schemas.prototype.resolve = function (id) {
  if (this.store[id] === undefined) {
    const error = new Error(`Schema with id '${id}' does not exist!`)
    error.code = 404
    throw error
  }
  return this.store[id]
}
```

File: test/internals/decorator.test.js [hunk 3]
```diff
@@ -48,18 +48,18 @@ test('checkExistence should find the instance if not given', t => {
...
+  server.add('test', () => { })
   t.ok(server.check('test'))
 })
 
 test('checkExistence should check the prototype as well', t => {
   t.plan(1)
-  function Instance () {}
-  Instance.prototype.test = () => {}
+  function Instance () { }
+  Instance.prototype.test = () => { }
```
[the remark this answers is not in the document]
Comment (mcollina) [reply test/internals/decorator.test.js]: I think vscode formatter, standard --fix is run every time.

**Comment by mcollina** [comment]
Very good so far. What is left to do? Docs for certain.

**Comment by cemremengu** [comment]
@StarpTech @mcollina I think this should be good to go after some final polishing (if necessary). 

Let me know if anything else needs to be done.

File: docs/Errros.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
+<h1 align="center">Fastify</h1>
+
+<a id="fastify-error-codes"></a>
+## Fastify Error Codes
+
+<a id="FST_ERR_CTP_ALREADY_PRESENT"></a>
+### FST_ERR_CTP_ALREADY_PRESENT
+
+The parser for this type was already registered.
```
Comment (mcollina) [inline docs/Errros.md]: type -> "content type"

File: docs/Errros.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+<a id="FST_ERR_CTP_EMPTY_TYPE"></a>
+### FST_ERR_CTP_EMPTY_TYPE
+
+The content type cannot be an empty string.
+
+<a id="FST_ERR_CTP_INVALID_HANDLER"></a>
+### FST_ERR_CTP_INVALID_HANDLER
+
+The content type handler should be a function.
```
Comment (mcollina) [inline docs/Errros.md]: I would rephrase it to "An invalid handler was passed ..."

**Comment by mcollina** [comment]
This PR needs to be merged by hand to keep the dual authorship of the work.

**Comment by cemremengu** [comment]
@mcollina I personally don't mind and wouldn't want anyone to waste time on that. All credits goes to @StarpTech, he did most of the work anyways.

File: docs/Errros.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+<a id="FST_ERR_SCH_MISSING_ID"></a>
+### FST_ERR_SCH_MISSING_ID
+
+The schema provided does not have `$id` property.
+
+<a id="FST_ERR_SCH_ALREADY_PRESENT"></a>
+### FST_ERR_SCH_ALREADY_PRESENT
+
+A schema with the same `id` already exists.
```
Comment (StarpTech) [inline docs/Errros.md]: Be consistent use `id` or `$id`.

File: docs/Errros.md [hunk 7]
```diff
@@ -66,12 +66,12 @@ The hook callback must be a function.
...
 ### FST_ERR_LOG_INVALID_DESTINATION
 
-Logger acceptes either a `stream` or a `file` as the destination.
+Logger acceptes either a `'stream'` or a `'file'` as the destination.
 
 <a id="FST_ERR_REP_INVALID_PAYLOAD_TYPE"></a>
 ### FST_ERR_REP_INVALID_PAYLOAD_TYPE
 
-Reply payload can either be a `string` or a `Buffer`.
+Reply payload can either be a `'string'` or a `'Buffer'`.
```
Comment (StarpTech) [inline docs/Errros.md]: You mean without quotes right? 😄

Comment (cemremengu) [reply docs/Errros.md]: Hang on almost there 😄

**Comment by StarpTech** [comment]
Ready to merge?

**Comment by mcollina** [comment]
go for it!

Il giorno ven 12 ott 2018 alle 12:37 Dustin Deus <notifications@github.com>
ha scritto:

**Comment by StarpTech** [comment]
Thank you @cemremengu for takeover.
